```{r setup, include=F}
library("optparse");
library("hwriter");
library("VariantAnnotation");
library("BSgenome.Hsapiens.UCSC.hg19");
library("BSgenome.Mmusculus.UCSC.mm10");
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library("reshape")
library("boot")
library('plyr')
library('dplyr')
library("ggplot2")
library("RColorBrewer")
require(gplots)
library("gridExtra")
require(NMF)
require(lsa)
require(reshape2)
library('RMySQL')

optList <- list(
                make_option("--name", default = '', type = "character", action = "store", help = "report name"),
                make_option("--alexandrovData", default = '~/share/reference/Alexandrov_NMF_signatures.txt', type = "character", action = "store", help = "alexandrov nmf signatures"),
                make_option("--lowPlus", default = F, help = "use only positions with effects annotated as greater than LOW"),
                make_option("--novel", default = T, help = "use only positions with GMAF < 0.01 or no GMAF"),
                make_option("--genome", default = 'b37', help = "reference genome"),
                make_option("--outDir", default = NULL, type = "character", action = "store", help = "output directory"),
                make_option("--targetBed", default = NULL, type = "character", action = "store", help = "target intervals in bed format"))

parser <- OptionParser(usage = "%prog [options] [vcf file(s)]", option_list = optList);
arguments <- parse_args(parser, positional_arguments = T, args = args);
opt <- arguments$options;

if (length(arguments$args) < 1) {
    cat("Need vcf file(s)\n");
    print_help(parser);
    stop();
}

vcfFiles <- arguments$args
outFile <- opt$outFile
if (opt$genome == "b37" || opt$genome == "hg19") {
    genome <- BSgenome.Hsapiens.UCSC.hg19
    txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
    genomeName <- 'hg19'
} else if (opt$genome == "mm10" || opt$genome == "GRCm38") {
    genome <- BSgenome.Mmusculus.UCSC.mm10
    txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
    genomeName <- 'mm10'
}

txdb <- renameSeqlevels(txdb, sub('chr', '', seqlevels(txdb)))
txByGenes <- transcriptsBy(txdb, 'gene')
seqlevels(genome) <- sub('chr', '', seqlevels(genome))
bases <- c("A", "C", "G", "T")

theme_set(theme_bw(base_size = 18))
```

# `r opt$name` Mutational Signature Report
---
### Raymond Lim

```{r loadAlexandrov, cache=T}
alexandrov <- read.table(opt$alexandrovData, sep = '\t', header = T)
sigs <- alexandrov[, grepl('Signature', colnames(alexandrov))]
```


```{r trinucleotideFreqs, cache=F}
genSeq <- getSeq(genome)
trintFq <- trinucleotideFrequency(genSeq)
trintFq <- colSums(trintFq) / sum(as.numeric(trintFq))
ratio <- rep(1.0, length(trintFq))
names(ratio) <- names(trintFq)
if (!is.null(opt$targetBed)) {
    bed <- import(opt$targetBed)
    targetSeq <- getSeq(genome, bed)
    targetTrintFq <- trinucleotideFrequency(targetSeq)
    targetTrintFq <- colSums(targetTrintFq) / sum(as.numeric(targetTrintFq))
    ratio <- trintFq / targetTrintFq
}
```

```{r loadvcf, include=F, cache=F}
vcfs <- list()
for (vcfFile in vcfFiles) {
    s <- sub('\\..*', '', vcfFile)
    s <- sub('.*/', '', s)

    temp <- tempfile()
    zipped <- bgzip(vcfFile, temp)
    idx <- indexTabix(temp, "vcf")
    cat('done\n')

    tab <- TabixFile(zipped, idx)
    open(tab)

    i <- 1
    vcf <- readVcf(tab, genomeName)
    if (nrow(vcf) > 0 && opt$lowPlus) {
        x <- sapply(info(vcf)$ANN, function(x) any(grepl('MODERATE|LOW|HIGH',
                                                         x)))
        vcf <- vcf[x,]
    }
    if (nrow(vcf) > 0 && opt$novel) {
        x <- is.na(info(vcf)$GMAF) | info(vcf)$GMAF < 0.01
        vcf <- vcf[x,]
    }
    if (nrow(vcf) > 0) {
        seqlevels(vcf) <- sub('chr', '', seqlevels(vcf))
        vcf <- vcf[sapply(rowRanges(vcf)$ALT, length) == 1]
        rowRanges(vcf)$MUT <- paste(rowRanges(vcf)$REF, unlist(rowRanges(vcf)$ALT), sep = ">")
        # query transcript ids
        ol <- findOverlaps(rowRanges(vcf), txByGenes)
        subjectStrands <- sapply(txByGenes[subjectHits(ol)], function(x) paste(unique(as.character(strand(x))), collapse = ','))
        queryStrands <- tapply(subjectStrands, queryHits(ol), function(x) paste(unique(x), collapse = ","))
        info(vcf)$strand <- NA
        info(vcf)$strand[as.integer(names(queryStrands))] <- queryStrands
        info(vcf)$transcribed <- F
        info(vcf)$transcribed[is.na(info(vcf)$strand)] <- NA
        info(vcf)$transcribed[rowRanges(vcf)$MUT %in% c("G>A", "G>C", "G>T", "A>C", "A>G", "A>T") & grepl('\\+', info(vcf)$strand)] <- T
        info(vcf)$transcribed[rowRanges(vcf)$MUT %in% c("C>A", "C>G", "C>T", "T>A", "T>C", "T>G") & grepl('-', info(vcf)$strand)] <- T

        rowRanges(vcf)$MUT[rowRanges(vcf)$MUT == "G>T"] <- "C>A"
        rowRanges(vcf)$MUT[rowRanges(vcf)$MUT == "G>C"] <- "C>G"
        rowRanges(vcf)$MUT[rowRanges(vcf)$MUT == "G>A"] <- "C>T"
        rowRanges(vcf)$MUT[rowRanges(vcf)$MUT == "A>T"] <- "T>A"
        rowRanges(vcf)$MUT[rowRanges(vcf)$MUT == "A>G"] <- "T>C"
        rowRanges(vcf)$MUT[rowRanges(vcf)$MUT == "A>C"] <- "T>G"
        rowRanges(vcf)$MUT <- factor(rowRanges(vcf)$MUT, levels = c("C>A", "C>G", "C>T", "T>A", "T>C", "T>G"))
        vcfs[[s]] <- vcf
    }
}
```

```{r pie, eval = F, echo = F}
for (s in names(vcfs)) {
    vcf <- vcfs[[s]]
    cols <- c("C>A" = "lightblue", "C>G" = "black", "C>T" = "red", "T>A" = "grey", "T>C" = "lightgreen",  "T>G" = "pink")
    main <- paste(s, " (n=", length(rowRanges(vcf)), ")", sep = '')
    pie(table(rowRanges(vcf)$MUT), col = cols, main = main)
    if (sum(info(vcf)$transcribed, na.rm = T) > 0 && sum(!info(vcf)$transcribed, na.rm = T) > 0) {
        main <- paste('transcribed', s, " (n=", sum(info(vcf)$transcribed, na.rm = T), ")", sep = '')
        pie(table(rowRanges(subset(vcf, info(vcf)$transcribed))$MUT), col = cols, main = main)
        main <- paste('untranscribed', s, " (n=", sum(!info(vcf)$transcribed, na.rm = T), ")", sep = '')
        pie(table(rowRanges(subset(vcf, !info(vcf)$transcribed))$MUT), col = cols, main = paste('untranscribed', main))
    }
}
```


```{r mutCountPlots, fig.height = 7, fig.width=28}
plotMutBarplot <- function(X, tit) {
    cols <- c("C>A" = "lightblue", "C>G" = "black", "C>T" = "red", "T>A" = "grey", "T>C" = "lightgreen",  "T>G" = "pink")
    tit <- paste(tit, ' (n = ', sum(X$value, na.rm = T), ")", sep = '')
    ggplot(X, aes(x = factor(Trinucleotide), y = value, fill = Substitution.Type)) +
        geom_bar(stat = 'identity') + facet_grid(~ Substitution.Type, scales = 'free_x') +
        xlab("") + ylab("") + ggtitle(tit) +
        theme(axis.text.x = element_text(angle = 90, hjust = 1), legend.position = 'none') +
        scale_fill_manual(values = cols)
}

plotMutBarplotStranded <- function(X, tit) {
    cols <- c('blue', 'red')
    tit <- paste(tit, ' (n = ', sum(X$value, na.rm = T), ")", sep = '')
    ggplot(X, aes(x = factor(Trinucleotide), y = value, fill = transcribed)) +
        geom_bar(stat = 'identity', position = 'dodge') + facet_grid(~ Substitution.Type, scales = 'free_x') +
        theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
        scale_fill_manual(name = "", values = cols, labels = c("Transcribed strand", 'Untranscribed strand')) + 
        xlab("") + ylab("") + ggtitle(tit)
}

plotMutPiechart <- function(X) {
    cols <- c("C>A" = "lightblue", "C>G" = "black", "C>T" = "red", "T>A" = "grey", "T>C" = "lightgreen",  "T>G" = "pink")
    X %>% group_by(Substitution.Type) %>% summarise(value = sum(value)) %>%
        ggplot(aes(x = "", y = value, fill = Substitution.Type)) +
        geom_bar(width = 1, stat = 'identity') + scale_fill_manual(values = cols) + 
        coord_polar("y") + xlab("") + ylab("")
}

createMutationTable <- function(vcf) {
    svcf <- split(rowRanges(vcf), rowRanges(vcf)$MUT)
    tabs <- list()
    for (mut in names(svcf)) {
        if (length(svcf[[mut]]) > 0) {
            x <- resize(flank(svcf[[mut]], width = 1, start = T, both = T), width = 3)
            genome(x) <- genome(genome)[1]
            seqs <- getSeq(genome, x)
            if (any(as.character(subseq(seqs, 2, 2)) == "G")) {
                seqs[as.character(subseq(seqs,2,2)) == "G"] <- reverseComplement(seqs[as.character(subseq(seqs,2,2)) == "G"])
            }
            if (any(as.character(subseq(seqs, 2, 2)) == "A")) {
                seqs[as.character(subseq(seqs, 2, 2)) == "A"] <- reverseComplement(seqs[as.character(subseq(seqs,2,2)) == "A"])
            }
            x <- as.character(subseq(seqs[1], 2, 2))
            lvls <- paste(rep(bases, each = 4), x, bases, sep = '')
            tabs[[mut]] <- table(factor(as.character(seqs), level = lvls))
        }
    }
    tab <- melt(tabs)
    X <- cbind(alexandrov[, 1:2], value = 0)
    m <- match(paste(tab$L1, tab$Var.1), paste(X$Substitution.Type, X$Trinucleotide))
    X$value[m] <- tab$value
    X
}

for (s in names(vcfs)) {
    vcf <- vcfs[[s]]
    X <- createMutationTable(vcf)
    p1 <- plotMutBarplot(X, s)
    p2 <- plotMutPiechart(X)
    grid.arrange(p1, p2, ncol = 2, widths = c(4, 2))

    Y <- transform(X, value = value * ratio[as.character(Trinucleotide)])
    Y <- transform(Y, value = value / sum(value))

    p1 <- plotMutBarplot(Y, paste('normalized', s))
    p2 <- plotMutPiechart(Y)
    grid.arrange(p1, p2, ncol = 2, widths = c(4, 2))
}

for (s in names(vcfs)) {
    vcf <- vcfs[[s]]
    if (sum(info(vcf)$transcribed, na.rm = T) > 0 && sum(!info(vcf)$transcribed, na.rm = T) > 0) {
        transcribedTab <- cbind(createMutationTable(subset(vcf, info(vcf)$transcribed)), transcribed = T)
        untranscribedTab <- cbind(createMutationTable(subset(vcf, !info(vcf)$transcribed)), transcribed = F)
        X <- rbind(transcribedTab, untranscribedTab)
        p1 <- plotMutBarplotStranded(X, s)
        p2 <- plotMutPiechart(transcribedTab)
        p3 <- plotMutPiechart(untranscribedTab)
        lom <- matrix(c(1,1, 2,3), nrow = 2, ncol = 2)
        grid.arrange(p1, p2, p3, layout_matrix = lom,  widths = c(4, 2))

        Y <- X %>% group_by(transcribed) %>% transform(value = value * ratio[as.character(Trinucleotide)])
        Y <- transform(Y, value = value / sum(value))

        p1 <- plotMutBarplotStranded(Y, paste('normalized', s))
        p2 <- plotMutPiechart(subset(Y, transcribed))
        p3 <- plotMutPiechart(subset(Y, !transcribed))
        grid.arrange(p1, p2, p3, layout_matrix = lom,  widths = c(4, 2))
    }
}

```

```{r bootPlot, fig.width=12}
bootFun <- function(x) {
    nval <- x$value * ratio[x$Trinucleotide]
    nval <- nval / sum(nval)
    sigs <- x[, grepl("Signature", colnames(x))]
    apply(sigs, 2, function(x) cor(nval, x))
}

ranFun <- function(p, d) {
    s <- sample.int(nrow(p), size = sum(p$value), replace = T, prob = p$value / sum(p$value))
    y <- melt(table(p[s,c(1,2)]))
    m <- match(paste(y$Substitution.Type, y$Trinucleotide), paste(p$Substitution.Type, p$Trinucleotide))
    p[m[!is.na(m)], "value"] <- y$value[!is.na(m)]
    p
}

for (s in names(vcfs)) {
    X <- inner_join(alexandrov, createMutationTable(vcfs[[s]]))
    boots <- boot(X, bootFun, R = 1000, ran.gen = ranFun, sim = 'parametric')
    boots.sd <- apply(boots$t, 2, sd)
    ci <- norm.ci(boots, index = 1:ncol(sigs))
    cols <- ifelse(boots$t0 > ci[,2] & boots$t0 < ci[,3], 'grey', 'red')
    n <- sub('Signature.', '', colnames(sigs))
    par(mfrow = c(2,1), mar = c(3,5,3,3))
    barCenters <- barplot(boots$t0, ylim = c(min(boots$t - boots.sd), max(boots$t + boots.sd)), names.arg = n, col = cols, main = s, ylab = 'Correlation')
    segments(barCenters, boots$t0 - boots.sd, barCenters, boots$t0+boots.sd, lwd = 1)
    # vote barplot
    barplot(table(factor(n[apply(boots$t, 1, which.max)], levels = n)), ylab = '# Votes')
}
```

```{r nmf, fig.width = 10, fig.height = 10}
rownames(alexandrov) <- with(alexandrov, gsub(">", ".", paste(Trinucleotide, Substitution.Type, sep=".")))
alexandrovM <- data.matrix(alexandrov[,4:ncol(alexandrov)])

solveNMF <- function(x, inmatrix){
    coef <- fcnnls(x, inmatrix[rownames(x),, drop = F]) # reorder the rownames of the in matrix 
    colsum <- apply(coef$x, 2, sum)
    coef_x_scaled <- scale(coef$x, center=F, scale=colsum)
    return(coef_x_scaled)
}


tabs <- list()
for (s in names(vcfs)) {
    tabs[[s]] <- createMutationTable(vcfs[[s]])
    names(tabs[[s]])[3] <- s
}
X <- as.matrix(join_all(tabs)[,-(1:2)])
rownames(X) <- with(alexandrov, gsub(">", ".", paste(Trinucleotide, Substitution.Type, sep=".")))

nmfMutCounts <- solveNMF(alexandrovM, X)
colnames(nmfMutCounts) <- names(vcfs)

if (!is.null(opt$outDir)) {
    fn <- paste(opt$outDir, "/nmf_mutsig.txt", sep = "")
    write.table(nmfMutCounts, file = fn, quote = F, sep = '\t')
}

cols <- brewer.pal(9, 'Blues')
if (ncol(nmfMutCounts) > 2) {
    heatmap.2(nmfMutCounts, trace = 'none', margins = c(13, 8), cexCol = 0.8, col = cols, Rowv = F, Colv = F, dendrogram = 'none')
}

par(mar = c(10,5,5,5))
for (s in colnames(nmfMutCounts)) {
    barplot(nmfMutCounts[,s], main = s, las = 2)
}
```
